
# Guided Project: Sales Performance Analysis with SQL
## Learn to combine SQL's skills we've learned to perform practical data analysis
## Dataquest

In [1]:
%load_ext sql
%sql sqlite://
%sql attach "sales.db" as sales

 * sqlite://
Done.


[]

# Guided Project Introduction

In [2]:
%%sql

-- Title: Sales Performance Analysis
-- Description: we are trying to analyze the sales.db databases to find meanningful trends using SQL.
-- Goal: Our goal is to find the share of sales of the top three regions, the frequently bought products in these top three countrie and the most suitable days for intense marketing and promotion?

-- table_name table's content.
SELECT *
  FROM table_name;
    

 * sqlite://
Done.


[(1, 'products', 'Items in the stores'),
 (2, 'regions', 'Geographic areas where stores exist'),
 (3, 'sales', 'Transactions in the store alongside with product, region, and time dimensions'),
 (4, 'sales_details', 'Transactions in the stores with several details'),
 (5, 'table_name', 'Tables in the database'),
 (6, 'time', 'Dates')]

# Getting to know the database

In [3]:
%%sql

-- the first three rows of each table from our database

SELECT *
  FROM products
 LIMIT 3;

 * sqlite://
Done.


[(1, 'Puff Pastry - Slab', None),
 (2, 'Lamb - Loin Chops', None),
 (3, 'Flour - Fast / Rapid', None)]

In [4]:
%%sql

SELECT *
  FROM regions
 LIMIT 3;

 * sqlite://
Done.


[(1, 'Indonesia', None), (2, 'Portugal', None), (3, 'Netherlands', None)]

In [5]:
%%sql

SELECT *
  FROM sales
 LIMIT 3;

 * sqlite://
Done.


[(4, 868, 227, 154, -39864),
 (5, 436, 684, 724, 79272),
 (6, 529, 176, 478, -33632)]

In [6]:
%%sql

SELECT *
  FROM sales_details
 LIMIT 3;

 * sqlite://
Done.


[(868, '2013-06-26 13:50:36', 2013, 6, 26, 13, 50, 36, 4, 'Chickensplit Half', 'Pakistan', -39864),
 (436, '2012-06-14 09:34:24', 2012, 6, 14, 9, 34, 24, 5, 'Soup - Beef Conomme, Dry', 'Poland', 79272),
 (529, '2015-08-12 02:58:16', 2015, 8, 12, 2, 58, 16, 4, 'Tea - Earl Grey', 'China', -33632)]

In [7]:
%%sql

SELECT *
  FROM time
 LIMIT 3;

 * sqlite://
Done.


[(1, '2017-10-11 19:52:57', 2017, 10, 11, 19, 52, 57, 4),
 (2, '2018-01-24 19:15:24', 2018, 1, 24, 19, 15, 24, 4),
 (3, '2017-12-06 19:45:58', 2017, 12, 6, 19, 45, 58, 4)]


# Identifying the Tables We Need

In [8]:
%%sql

-- Selecting the identified table.

SELECT 'sales_details' AS selected_table;


 * sqlite://
Done.


[('sales_details',)]

In [9]:
%%sql
-- Counting the rows of the selected table.

SELECT COUNT(*) AS num_rows
  FROM sales_details;

 * sqlite://
Done.


[(1000,)]


# Filtering interesting rows


In [10]:
%%sql

-- Filtering the rows related to positive values from the `transaction_amount` column.

SELECT *
  FROM sales_details
 WHERE transaction_amount >= 0 
 LIMIT 3;

 * sqlite://
Done.


[(436, '2012-06-14 09:34:24', 2012, 6, 14, 9, 34, 24, 5, 'Soup - Beef Conomme, Dry', 'Poland', 79272),
 (146, '2015-02-06 12:18:18', 2015, 2, 6, 12, 18, 18, 6, 'Wine - Spumante Bambino White', 'France', 87806),
 (483, '2012-12-01 23:08:12', 2012, 12, 1, 23, 8, 12, 7, 'Ketchup - Tomato', 'Kazakhstan', 98243)]

In [11]:
%%sql

-- Counting the positive values from the `transaction_amount` column compared to the number of rows in the filtered table.

SELECT COUNT(transaction_amount) AS num_positive_transactions, COUNT(*) AS num_rows
  FROM sales_details
 WHERE transaction_amount >= 0;

 * sqlite://
Done.


[(494, 494)]

3. Since the `num_positive_transactions` and `num_rows` are the same, we can conclude that there is no missing values in the `transaction_amount` column.


# Total Transaction Amount for Each Region


In [12]:
%%sql

-- Computing the total positive transaction amount for each region using the `GROUP BY` clause.

SELECT region_name, SUM(transaction_amount) AS total_transaction_amount
  FROM sales_details
 WHERE transaction_amount >= 0
 GROUP BY region_name;

 * sqlite://
Done.


[('Afghanistan', 131159),
 ('Albania', 235171),
 ('Argentina', 340583),
 ('Armenia', 77199),
 ('Australia', 187963),
 ('Azerbaijan', 159200),
 ('Bahamas', 148728),
 ('Bangladesh', 154916),
 ('Belarus', 25315),
 ('Benin', 34352),
 ('Bolivia', 10961),
 ('Brazil', 488294),
 ('Burkina Faso', 38329),
 ('Cameroon', 44091),
 ('Canada', 125266),
 ('China', 3587812),
 ('Colombia', 356547),
 ('Croatia', 140920),
 ('Cuba', 196686),
 ('Cyprus', 577),
 ('Czech Republic', 94021),
 ('Denmark', 66339),
 ('Dominican Republic', 126001),
 ('Ecuador', 26411),
 ('Egypt', 59730),
 ('Estonia', 32547),
 ('Ethiopia', 276368),
 ('Finland', 196962),
 ('France', 871335),
 ('Georgia', 90044),
 ('Germany', 74158),
 ('Greece', 346326),
 ('Guatemala', 9567),
 ('Guinea', 49896),
 ('Guyana', 124443),
 ('Honduras', 169520),
 ('Hungary', 80988),
 ('Indonesia', 2565442),
 ('Iran', 353566),
 ('Iraq', 96055),
 ('Ireland', 89760),
 ('Israel', 81269),
 ('Ivory Coast', 142272),
 ('Japan', 138545),
 ('Kazakhstan', 121368),
 ('Kyrgyzstan', 95798),
 ('Laos', 119385),
 ('Latvia', 74817),
 ('Libya', 81214),
 ('Luxembourg', 70793),
 ('Macedonia', 46793),
 ('Madagascar', 92723),
 ('Malawi', 122046),
 ('Mali', 85633),
 ('Mauritius', 42660),
 ('Mexico', 338072),
 ('Mongolia', 6955),
 ('Morocco', 97271),
 ('Netherlands', 87392),
 ('New Caledonia', 31718),
 ('Nicaragua', 54865),
 ('Nigeria', 150886),
 ('Norway', 214806),
 ('Pakistan', 121607),
 ('Palestinian Territory', 291565),
 ('Peru', 190533),
 ('Philippines', 1097122),
 ('Poland', 1342022),
 ('Portugal', 559054),
 ('Russia', 1053683),
 ('Saudi Arabia', 28475),
 ('Serbia', 95519),
 ('Sierra Leone', 41685),
 ('Slovenia', 47833),
 ('Somalia', 66332),
 ('South Korea', 229805),
 ('Spain', 355871),
 ('Sweden', 777801),
 ('Switzerland', 131534),
 ('Syria', 123207),
 ('Tajikistan', 58541),
 ('Tanzania', 245208),
 ('Thailand', 413431),
 ('Togo', 201465),
 ('Turkmenistan', 47280),
 ('Uganda', 190927),
 ('Ukraine', 228561),
 ('United Kingdom', 74921),
 ('United States', 1005656),
 ('Uzbekistan', 68969),
 ('Venezuela', 112740),
 ('Vietnam', 259546),
 ('Yemen', 22777)]


# Top Three Regions


In [13]:
%%sql


-- Finding the top three regions in relation to the total positive transaction amount.

SELECT region_name, SUM(transaction_amount) AS total_transaction_amount
  FROM sales_details
 WHERE transaction_amount >= 0
 GROUP BY region_name
 ORDER BY total_transaction_amount DESC
 LIMIT 3;


 * sqlite://
Done.


[('China', 3587812), ('Indonesia', 2565442), ('Poland', 1342022)]

# Share of Sales


In [14]:
%%sql

-- Computes the share of sales of the top three regions in relation to the total positive transaction amount.

SELECT region_name, SUM(transaction_amount) * 100.0 / 24064499 AS total_transaction_amount_props
  FROM sales_details
 WHERE transaction_amount >= 0
 GROUP BY region_name
 ORDER BY total_transaction_amount_props DESC
 LIMIT 3;
    


 * sqlite://
Done.


[('China', 14.909148950077872),
 ('Indonesia', 10.660691502449314),
 ('Poland', 5.576770993653348)]

In [15]:
%%sql

-- Summing the share of sales
SELECT ROUND(14.909148950077872, 2) + ROUND(10.660691502449314, 2) + ROUND(5.576770993653348, 2) AS top_3_share_market

 * sqlite://
Done.


[(31.15,)]

3. The highest sales in our store come from Asian countries. Our customers are probably many in this part of the world.


# Finding Frequently Bought Products


In [16]:
%%sql


-- Finding the frequently bought products in China, Indonesia, and Poland.

SELECT region_name, product_name, COUNT(*) AS num_transactions, AVG(transaction_amount) AS avg_transaction_amount
  FROM sales_details
 WHERE transaction_amount >= 0 AND (region_name = 'China'  OR region_name = 'Indonesia'  OR region_name = 'Poland')
 GROUP BY region_name, product_name
HAVING COUNT(*) > 1
 ORDER BY num_transactions DESC, region_name, product_name;

 * sqlite://
Done.


[('Indonesia', 'Beef - Rib Eye Aaa', 3, 60284.333333333336),
 ('China', 'Coke - Diet, 355 Ml', 2, 46720.0),
 ('Indonesia', 'Cheese - La Sauvagine', 2, 54260.5),
 ('Indonesia', 'Fennel - Seeds', 2, 36930.0),
 ('Indonesia', 'Oneshot Automatic Soap System', 2, 58257.0),
 ('Indonesia', 'Sprouts - Baby Pea Tendrils', 2, 6054.5)]

3. For example, in China, we always have to ensure to have the `Coke - Diet, 355 Ml` in stock and in Indonesia, `Beef - Rib Eye Aaa`. Notice that, Poland is not in the top 3 sales because some products are bought very frequently. Maybe the products there are just expensive.


# Finding Promotion Days


In [17]:
%%sql

-- Finding the days of the week where the number of transactions is greater or equal to `10` in China, Indonesia, and Poland.
   
SELECT region_name, time_day_of_week, COUNT(*) AS num_transactions, SUM( transaction_amount) as total_transaction_amount
  FROM sales_details
 WHERE transaction_amount >= 0 AND (region_name = 'China'  OR region_name = 'Indonesia'  OR region_name = 'Poland' )
 GROUP BY  region_name, time_day_of_week
HAVING  COUNT(*) >= 10
 ORDER BY region_name, time_day_of_week;

 * sqlite://
Done.


[('China', 3, 14, 687399),
 ('China', 4, 14, 578406),
 ('China', 5, 13, 607495),
 ('China', 7, 14, 702672),
 ('Indonesia', 5, 13, 717903)]

2. In China most of the days are interesting while in Indonesia it is only on Friday. 